### DISCUSSION TITLE
Изменение Состояния как минимальная структурная единица игры

### DISCUSSION THREAD

**Discussion by nergal-perm** [nergal-perm/tg-core#33] [Закрытые обсуждения] [answered]
Любая настольная пошаговая игра - это не что иное, как последовательное изменение состояния - игроки перемещают свои фишки, перекладывают жетоны, берут в руку и выкладывают на стол карты - все это можно представить в виде изменения состояния Игры. Важно, что изменения не происходят одновременно и параллельно - все они строго последовательны.

Исходя из этого можно выделить минимальную неделимую часть игры - атомарное `Изменение состояния`. Наверное, можно представить их в виде перемещения чего-либо между двумя частями состояния игры в целом. Примеры с картами и разными фишками однозначно транслируются в такую форму, но, например, попытки отгадать число в "Быках-коровах" - уже с натяжкой. Тем не менее, стоит над этим подумать. 

Однако с точки зрения игры отдельные изменения состояния не особенно интересны. Гораздо важнее то, что они происходят в результате какого-то `Действия`, предпринимаемого игроком или игрой. Именно `Действие` является минимальной единицей игры, имеющей смысл. В результате `Действий` изменяется состояние, после совершения `Действия` игра может передать ход другому игроку или еще как-то изменить свою фазу (начать новый раунд, оценить условия завершения игры и т.п.). 

Мы как-то говорили о том, что любое состояние любой игры можно представить в виде результата последовательного применения ряда каких-то входящих команд, и вот `Действие`, мне кажется, подходит на эту роль.

Что важно для Действия?

- Всегда есть один и только один Инициатор - либо игрок, либо игра.
- Действие без изменения состояния лишено смысла.
- С действием может быть связано любое количество изменений состояния (одно или больше), в том числе и изменение фазы игры (смена активного игрока, и т.п.), то есть выполнение действия может продвинуть игру дальше по структуре ходов и раундов.
- Фактическое содержание действия (то есть реальные изменения состояния) специфичны для каждой игры, но само поведение действия (применение этих изменений и оповещение игры о совершении действия) одинаково для всех игр.

**Comment by citerm** [comment]
нечего добавить.

**Comment by nergal-perm** [comment, marked as answer]
Перенес в [wiki](https://github.com/nergal-perm/tg-core/wiki/%D0%94%D0%B5%D0%B9%D1%81%D1%82%D0%B2%D0%B8%D0%B5-%D0%BA%D0%B0%D0%BA-%D1%81%D1%82%D1%80%D1%83%D0%BA%D1%82%D1%83%D1%80%D0%BD%D0%B0%D1%8F-%D0%B5%D0%B4%D0%B8%D0%BD%D0%B8%D1%86%D0%B0)
